stix2/sources/__init__.py

```diff
@@ -11,104 +11,106 @@
 
         Args:
             stix_id (str): the id of the STIX object to retrieve.
-            allow_custom (bool): whether to retrieve custom objects/properties
-                or not. Default: False.
 
         Returns:
             stix_obj: the single most recent version of the STIX
                 object specified by the "id".
 
         """
-        return self.source.get(stix_id, allow_custom=allow_custom)
+        return NotImplementedError()
 
-    def all_versions(self, stix_id, allow_custom=False):
+    @abstractmethod
+    def all_versions(self, stix_id):  # pragma: no cover
         """Retrieve all versions of a single STIX object by ID.
 
-        Implement: Translate all_versions() call to the appropriate DataSource call
+        Implement: Define a function that performs any custom behavior before
+            calling the associated DataSource all_versions() method.
 
         Args:
             stix_id (str): the id of the STIX object to retrieve.
-            allow_custom (bool): whether to retrieve custom objects/properties
-                or not. Default: False.
 
         Returns:
             stix_objs (list): a list of STIX objects
 
         """
-        return self.source.all_versions(stix_id, allow_custom=allow_custom)
+        return NotImplementedError()
 
-    def query(self, query=None, allow_custom=False):
+    @abstractmethod
+    def query(self, query=None):  # pragma: no cover
         """Retrieve STIX objects matching a set of filters.
 
         Implement: Specific data source API calls, processing,
         functionality required for retrieving query from the data source.
 
+        Define custom behavior before calling the associated DataSource query()
+
         Args:
             query (list): a list of filters (which collectively are the query)
                 to conduct search on.
-            allow_custom (bool): whether to retrieve custom objects/properties
-                or not. Default: False.
 
         Returns:
             stix_objs (list): a list of STIX objects
 
         """
-        return self.source.query(query=query)
+        return NotImplementedError()
 
-    def add(self, stix_objs, allow_custom=False):
-        """Store STIX objects.
+    @abstractmethod
+    def add(self, stix_objs):  # pragma: no cover
+        """Method for storing STIX objects.
 
-        Translates add() to the appropriate DataSink call.
+        Define custom behavior before storing STIX objects using the associated
+        DataSink. Translates add() to the appropriate DataSink call.
 
         Args:
             stix_objs (list): a list of STIX objects
-            allow_custom (bool): whether to allow custom objects/properties or
-                not. Default: False.
+
         """
-        return self.sink.add(stix_objs, allow_custom=allow_custom)
+        return NotImplementedError()
 
 
-class DataSink(object):
+class DataSink(with_metaclass(ABCMeta)):
```

Comment:
Having DataSink and DataSource as abstract classes is OK. I just think DataStore can provide a lot of default functionality (delegating to `source` and `sink` attributes.